From 758b887b1c69214f16d5f5c80c187fa6efb7d855 Mon Sep 17 00:00:00 2001
From: Caesar Wang <wxt@rock-chips.com>
Date: Fri, 3 Nov 2017 12:58:14 +0800
Subject: [PATCH] MALI: utgard: RK: fixes the mali driver load failed on rk3036

As the gpu driver used the devfreq thermal, The mali will be failure to
load the driver since the rk platform hadn't the driver requested
deferred probing.

as the following is failure log:
[    7.126149] Mali: ERR: drivers/gpu/arm/mali400/mali/linux/mali_kernel_linux.c
[    7.126157]            mali_probe() 550
                          mali_probe(): Failed to initialize platform device.
[    7.126191] mali-utgard: probe of 10091000.gpu failed with error -14
..

Add the SoCs judge if we will use the devfreq-thermal for mali driver.
If the other SoCs has the similar issue, we can do this way.

Change-Id: I9faf6530119adb99efeac491e665641ee8b1143d
Signed-off-by: Caesar Wang <wxt@rock-chips.com>
---
 drivers/gpu/arm/mali400/mali/linux/mali_devfreq.c | 3 +++
 drivers/gpu/arm/mali400/mali/platform/arm/arm.c   | 4 ++++
 drivers/gpu/arm/mali400/mali/platform/rk/rk.c     | 3 +++
 3 files changed, 10 insertions(+)

diff --git a/drivers/gpu/arm/mali400/mali/linux/mali_devfreq.c b/drivers/gpu/arm/mali400/mali/linux/mali_devfreq.c
index 239720043d99..3eac07d76766 100644
--- a/drivers/gpu/arm/mali400/mali/linux/mali_devfreq.c
+++ b/drivers/gpu/arm/mali400/mali/linux/mali_devfreq.c
@@ -262,6 +262,9 @@ int mali_devfreq_init(struct mali_device *mdev)
 	}
 
 #ifdef CONFIG_DEVFREQ_THERMAL
+	if (of_machine_is_compatible("rockchip,rk3036"))
+		return 0;
+
 	/* Initilization last_status it will be used when first power allocate called */
 	mdev->devfreq->last_status.current_frequency = mdev->current_freq;
 
diff --git a/drivers/gpu/arm/mali400/mali/platform/arm/arm.c b/drivers/gpu/arm/mali400/mali/platform/arm/arm.c
index 4e09aca4c6f0..bf7ffaff87e4 100755
--- a/drivers/gpu/arm/mali400/mali/platform/arm/arm.c
+++ b/drivers/gpu/arm/mali400/mali/platform/arm/arm.c
@@ -534,6 +534,10 @@ int mali_platform_device_init(struct platform_device *device)
 	}
 
 #if defined(CONFIG_MALI_DEVFREQ) && defined(CONFIG_DEVFREQ_THERMAL)
+	/* Some Socs didn't support the devfreq thermal for mali */
+	if (of_machine_is_compatible("rockchip,rk3036"))
+		return 0;
+
 	/* Get thermal zone */
 	gpu_tz = thermal_zone_get_zone_by_name("soc_thermal");
 	if (IS_ERR(gpu_tz)) {
diff --git a/drivers/gpu/arm/mali400/mali/platform/rk/rk.c b/drivers/gpu/arm/mali400/mali/platform/rk/rk.c
index 793fae630b09..f354d2cf94f0 100644
--- a/drivers/gpu/arm/mali400/mali/platform/rk/rk.c
+++ b/drivers/gpu/arm/mali400/mali/platform/rk/rk.c
@@ -531,6 +531,9 @@ int mali_platform_device_init(struct platform_device *pdev)
 	}
 
 #if defined(CONFIG_MALI_DEVFREQ) && defined(CONFIG_DEVFREQ_THERMAL)
+	if (of_machine_is_compatible("rockchip,rk3036"))
+		return 0;
+
 	err = power_model_simple_init(pdev);
 	if (err) {
 		E("fail to init simple_power_model, err : %d.", err);
-- 
2.35.3

